GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorWithIProductUpgraderTests.cs

```diff
@@ -0,0 +1,202 @@
+﻿using GVFS.Common;
+using GVFS.Common.Git;
+using GVFS.Tests.Should;
+using GVFS.UnitTests.Mock.Common;
+using GVFS.UnitTests.Windows.Mock.Upgrader;
+using GVFS.Upgrader;
+using Moq;
+using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+
+namespace GVFS.UnitTests.Upgrader
+{
+    [TestFixture]
+    public class UpgradeOrchestratorWithIProductUpgraderTests
+    {
+        private const string OlderThanLocalVersion = "1.0.17000.1";
+        private const string LocalGVFSVersion = "1.0.18115.1";
+        private const string NewerThanLocalVersion = "1.1.18115.1";
+
+        private delegate void TryGetNewerVersionCallback(out System.Version version, out string message);
+        private delegate void UpgradeAllowedCallback(out string message);
+        private delegate void TryRunPreUpgradeChecksCallback(out string delegateMessage);
+        private delegate void TryDownloadNewestVersionCallback(out string message);
+        private delegate void TryRunInstallerCallback(InstallActionWrapper installActionWrapper, out string error);
+
+        private MockTracer Tracer { get; set; }
+        private MockTextWriter Output { get; set; }
+        private MockInstallerPrerunChecker MockPreRunChecker { get; set; }
+        private Mock<LocalGVFSConfig> MockLocalConfig { get; set; }
+        private Mock<IProductUpgrader> MockUpgrader { get; set; }
+
+        private UpgradeOrchestrator orchestrator { get; set; }
+
+        [SetUp]
+        public void Setup()
+        {
+            this.Tracer = new MockTracer();
+            this.Output = new MockTextWriter();
+            this.MockPreRunChecker = new MockInstallerPrerunChecker(this.Tracer);
+            this.MockPreRunChecker.Reset();
+            this.MockUpgrader = this.DefaultUpgrader();
+            this.orchestrator = new UpgradeOrchestrator(
+                this.MockUpgrader.Object,
+                this.Tracer,
+                this.MockPreRunChecker,
+                input: null,
+                output: this.Output);
+        }
+
+        [TestCase]
+        public void UpgradesWhenUpgradeAvailable()
```

Comment:
Naming suggestion:  Include a more explicit description of what's being tested and the the expected result in the name.  

For example, rather than `UpgradesWhenUpgradeAvailable` try something like `OrchestratorExecuteSucceedsWhenUpgradeAvailable`.